Add `ExtensionType` trait and `CanonicalExtensionType` enum

mbrobbel opened the pull request:
- closes #4472?
 
# Rationale for this change
 
It would be nice to better support reading and writing the Arrow canonical `uuid` and `json` extension types with the arrow and parquet crate i.e. mapping between the arrow extension type and the parquet logical `uuid` and `json` types.

# What changes are included in this PR?

This adds an `ExtensionType` trait, some impls for canonical extension types and a `CanonicalExtensionType` enum for canonical extension types.

# Are there any user-facing changes?

Users can now annotate their fields with extension types, and for `uuid` and `json` they are propagated via the arrow writer to map to the parquet `uuid` and `json` logical types.

kylebarron commented:
Maybe ExtensionType could be a trait to be externally implementable and not limited to canonical extension types?

mbrobbel commented:
I haven't had time to work on this, but I'm planning to pick this up later.

alamb commented:
Thanks @mbrobbel  -- marking this PR as draft as I think it still has planned but not yet completed work

aykut-bozkurt commented:
Thanks for the PR. This seems very useful to support not yet mapped logical types. e.g. json

mbrobbel commented:
I updated the PR to define a trait for extension types instead. Ready for another round of feedback.

mbrobbel commented:
This is ready for another round of reviews.

I updated the trait and changed some field methods, added implementations for the current set of canonical extension types (behind a feature flag) and added more tests and docs. Maybe this PR is too big now, but I don't mind splitting it up if needed.

Roundtripping through parquet is not implemented in this PR.

alamb commented:
Thank you @mbrobbel  -- I plan to review this carefully tomorrow

alamb commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+///
+/// # Example
+///
+/// The example below demonstrates how to implement this trait for a `Uuid`
```
this is a really nice example and illustrates how this API works

alamb commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+    ///
+    /// If an extension type defines no metadata it should use `()` to indicate
+    /// this.
+    type Metadata;
```
I wonder what is the usecase for the Metadata indirection? 

For instance, the example above seems like it would be simpler if it simply serialized / deserialized `UuidVersion` directly

I think that would mean the API would look like
```rust
trait ExtensionType {
  // serialize self to a string
  fn serialize(&self) -> Option<String>;
  // deserialize into self
  fn deserialize(metadata: Option<&str>) -> Result<Self, ArrowError>;
}
```

jayzhan211 replied:
In Json, `serde_json::Value` is the metadata instead of String, with your design I think we are not able to store metadata other than String

alamb replied:
I do think any metadata needs to be *serialized* to a string (which is the same as what the PR does)

```rust
    fn serialize_metadata(&self) -> Option<String>;
```

I am just wondering why have a separate Rust type for Metadata

For eample, if I were doing this my first inclination would be to do

```rust
impl ExtensionType for MyAwesomeType {
    ...
    type Metadata = Self;
    ....
}

So then I could just serialize/deserialize instances of my awesome type

jayzhan211 replied:
I see. The only benefit is that we can have a separate clean struct IF the metadata becomes pretty complex. For usual case, simply se/de without `type Metadata` is enough.

mbrobbel replied:
This doesn't work for all extension types. For example the `FixedShapeTensor` and `VariableShapeTensor` canonical extension types can't be constructed from just the metadata, they also need the field datatype (which explains the `ExtensionType::try_new` signature).

alamb replied:
Ah, got it. Maybe we can add some comments explaining the rationale a bit more. 

Or maybe we can default the `Metadata` type to `Self`

```rust
type Metadata = Self;
```

To be clear we can do this as a follow on PR (or never). I don't think it is necessary for this PR

mbrobbel replied:
Unfortunately associated type defaults are not stabilized yet. I added a note in [b78e692](https://github.com/apache/arrow-rs/pull/5822/commits/b78e6925d76b35a739341f9a86cde91e58074ecc).

alamb commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+
+//! Extension types.
+
+#[cfg(feature = "canonical-extension-types")]
```
💯 for making this behind a feature flag

It would be nice to add this to the documentation: 
- https://github.com/apache/arrow-rs/tree/main/arrow#feature-flags

Perhaps we could also note it is "experimental" and say we reserve the right to make breaking changes between even minor version releases -- that would give us flexibilty to iterate in the short term if necessary

mbrobbel replied:
[069642f](https://github.com/apache/arrow-rs/pull/5822/commits/069642fde6af993dfb94d6082bddaf0933740f8b)

> Perhaps we could also note it is "experimental"

Do you want to add an `experimental` feature to the `arrow` crate (same as in the `parquet` crate)?
https://github.com/apache/arrow-rs/blob/16642149b75f0b18d8e689ad9e2111edf4e429a9/parquet/Cargo.toml#L112-L113

Or just a note in the module docs?

alamb replied:
I think adding a note in the module docs is good enough

mbrobbel replied:
Added a warning in 5fec56d8ff81d14902274bd45e9b41e634a9379e.

alamb commented on arrow-schema/src/extension/canonical/mod.rs:
```diff
@@ -0,0 +1,142 @@
+/// Canonical extension types.
+///
+/// <https://arrow.apache.org/docs/format/CanonicalExtensions.html#format-canonical-extensions>
+#[non_exhaustive]
```
👍  for non exhaustive so we can add new extension types without breaking API changes

alamb reviewed (approved):
😮 -- thank you SO much @mbrobbel  -- I looked at the structure of this PR and I really like it. It is really nicely done. I have some minor comments, but really I found this PR really well structured, commented and tested 🏅 

This also implements the consensus reached on https://github.com/apache/arrow-rs/issues/4472 as I understand it

I have some API design questions, but I also think this is good enough to merge as is and I would be happy to maintain the code as is. 

If there are no objections to this approach from other reviewers in theory I will find the time to review the tests a bit more closely before merging, but the ones I have looked at so far looks 👍 👍 👍 



@westonpace @kylebarron @tustvold @yukkit perhap you would have some time to review / look over this PR as you were quite active on https://github.com/apache/arrow-rs/issues/4472

@emilk, @danking @jleibs, @crepererum and @findepi as you seem to have mentioned this PR in the past

jayzhan211 reviewed (approved):
It looks pretty nice!

jayzhan211 commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+    ///
+    /// [`Field`]: crate::Field
+    /// [`Field::metadata`]: crate::Field::metadata
+    fn serialize_metadata(&self) -> Option<String>;
```
Probably `Result<String>` is enough, I think None and "" in this case doesn't matter

mbrobbel replied:
I can't think of a practical situation where serialization of the extension type metadata would error, but I'm fine with adding a `Result` here.

The `Option` is required to support extension types without metadata (they should return `None` here). Note that this is different from `""`, because in that case the extension metadata key is added to the field custom metadata, whereas for `None` this key is skipped.

So with support for serialization errors this would return: `Option<Result<String, ArrowError>>`.

mbrobbel replied:
Per https://arrow.apache.org/docs/format/Columnar.html#extension-types:

> Extension types may or may not use the 'ARROW:extension:metadata' field.

Which I interpret as; it should also work when this key is missing in the field metadata for extension types that don't require metadata.

mbrobbel replied:
> It's your call

I think this should be discussed with other contributors.

mbrobbel replied:
Yes, I can imagine. Thanks for the suggestion.

alamb replied:
Maybe we can add a note in the comments and suggest for maximum compatibility extension types should always return `Some("")`

mbrobbel replied:
This would be in conflict with the docs for `deserialize_metadata`: 

> An extension type that defines no metadata should expect `None` for the serialized metadata and return `Ok(())`.

> This function should return an error when **unexpected metadata is set (for extension types without metadata)**

We could add a note as you suggested for `serialize_metadata` and then modify the comments in `deserialize_metadata` to be suggestions only?

The `Uuid` canonical extension type requires no metadata, so if we change the recommendation here we should also change that implementation maybe?

kylebarron reviewed (commented):
Oops forgot to commit changes

mbrobbel commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -2163,4 +2211,52 @@ mod tests {
+            Some(LogicalType::Json)
+        );
+
+        // TODO: roundtrip
+        // let arrow_schema = parquet_to_arrow_schema(&parquet_schema, None)?;
+        // assert_eq!(
+        //     arrow_schema.field(0).try_extension_type::<Json>()?,
+        //     Json::default()
+        // );
```
Yes, this requires a bit more work to propagate correctly.

alamb replied:
- Filed https://github.com/apache/arrow-rs/issues/7063 to track

mbrobbel commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+///
+///     // We just return a reference to the Uuid version.
+///     fn metadata(&self) -> &Self::Metadata {
+///         &self.0
```
Yes, the current signature requires that. We could change it to return `Self::Metadata`?

The pattern I used for the canonical extension types with metadata applied to your example:

```rust
pub struct PointType(PointTypeMetadata);

pub struct PointTypeMetadata(CoordType, Dimension);

impl ExtensionType for PointType {
    const NAME: &'static str = "geoarrow.point";

    type Metadata = PointTypeMetadata;

    fn metadata(&self) -> &Self::Metadata {
        &self.0
    }
}
```

mbrobbel commented on arrow-schema/src/extension/mod.rs:
```diff
@@ -0,0 +1,252 @@
+    ///
+    /// [`Field`]: crate::Field
+    /// [`Field::metadata`]: crate::Field::metadata
+    fn serialize_metadata(&self) -> Option<String>;
```
Ah this is great, I'll take a better look at this tomorrow and figure out if the current definition works with GeoArrow. 
Note that in the spec extension types are defined for Field, not for Array.

mbrobbel replied:
It looks like all the pieces are there (in `geoarrow-rs`), so I guess it will be straightforward to add `ExtensionType` implementations that can be used with the added extension type related methods of `Field`.

Tpt commented on arrow-schema/src/extension/canonical/bool8.rs:
```diff
@@ -0,0 +1,143 @@
+impl ExtensionType for Bool8 {
+    const NAME: &'static str = "arrow.bool8";
+
+    type Metadata = &'static str;
```
Maybe bad idea: use `()` here. This way it's not possible to encode bad metadata value (any string that is not empty)

mbrobbel replied:
With the proposed definition of the trait, `()` would indicate no metadata and results in the metadata key to be in unset in the field custom metadata map. Note that there is no actual storage for the metadata in this extension type. Both serialization and deserialization make sure the metadata is always an empty string.

Tpt replied:
Ok! Missed that! Thanks!

Tpt commented on arrow-schema/src/extension/canonical/bool8.rs:
```diff
@@ -0,0 +1,143 @@
+                "" => Ok(""),
+                _ => Err(ArrowError::InvalidArgumentError(ERR.to_owned())),
+            },
+        )
```
slightly shorter:
```suggestion
        if metadata.map_or(false, str::is_empty) {
           Ok("")
        } else {
           Err(ArrowError::InvalidArgumentError("Bool8 extension type expects an empty string as metadata".into()))
        }
```

mbrobbel replied:
[757b041](https://github.com/apache/arrow-rs/pull/5822/commits/757b041ee1a80735d1314dc905e299b6034535da)

Tpt commented on arrow-schema/src/extension/canonical/json.rs:
```diff
@@ -0,0 +1,189 @@
+///
+/// This type does not have any parameters.
+///
+/// Metadata is either an empty string or a JSON string with an empty
```
Based on implementation I guess it's not "an empty string" but "a JSON string with an empty string" (ie `"\"\""` and not `""`)

mbrobbel replied:
Good catch! [29a94cb](https://github.com/apache/arrow-rs/pull/5822/commits/29a94cbed185515f4eb8cad362f1d2c242fd8eac)

mbrobbel replied:
I did consider doing this. The spec states (https://arrow.apache.org/docs/format/CanonicalExtensions.html#json):

> Description of the serialization:
> Metadata is either an empty string or a JSON string with an empty object. **In the future, additional fields may be added**, but they are not required to interpret the array.

So I figured, I'll support the future use-case, however, if/when in the future fields are added here it would actually be better to use a strongly typed definition. I like your suggestion of using `Empty` here and since the metadata content currently is not `pub`, if fields are added in the future, we can just rename and modify the definition of the inner metadata type without introducing a breaking change.

[c6f0443](https://github.com/apache/arrow-rs/pull/5822/commits/c6f0443c7cc6eff4d0765366166545d34ef91d69)

Tpt replied:
Great! It might be possible to also use a similar approach for the `Bool8` metadata.

mbrobbel replied:
There is currently no metadata storage for `Bool8`, and the difference is that there is no note about future additions to the metadata serialization:

> Description of the serialization:
> Metadata is an empty string.

So, if there are breaking changes in the spec for this type, it would have to be a breaking change here too anyway.

Tpt reviewed (commented):
Just passing by and had a lot of fun reading this code (thank you)!

Please just ignore my comments if you feel so.

alamb commented:
I think we should leave this PR open at least through the weekend to gather additional feedback. I think given the size/scope of it perhaps we can plan to merge it after I cut the release candidate for the next arrow-rs release (mid-late next week) if there are no other concerns / comments.

THank you again @mbrobbel  for pushing this along

alamb commented:
I'll merge this PR in the next day or two unless anyone else has any additional comments or wants more time to comment

alamb commented:
I merged up from main to run the CI once more before merging this

alamb commented:
Epic work -- can't wait to see how this works in the real world

alamb commented:
For some reason this PR isn't showing up in the release notes  🤔 
- https://github.com/apache/arrow-rs/pull/7110

It is in the release though

kylebarron commented on arrow-schema/src/field.rs:
```diff
@@ -350,6 +355,167 @@ impl Field {
+    ///
+    /// This calls [`Field::try_with_extension_type`] and panics when it
+    /// returns an error.
+    pub fn with_extension_type<E: ExtensionType>(mut self, extension_type: E) -> Self {
```
I'll just note it's a little weird to me that `with_extension_type` returns `Self` whereas `try_with_extension_type` returns `Result<()>`. I would've expected `try_with_extension_type` to return `Result<Self>`?

mbrobbel replied:
The reasoning was: you may still want to use your field when `try_with_extension_type` returns an error.
An alternative would be to return `Result<Self, (Self, Error)>`?

kylebarron replied:
I suppose that makes sense!

alamb replied:
I was thinking about how to update the documentation to capture this conversation but I can't really come up with an specific changes to recommend
